masq/src/terminal_interface.rs

```diff
@@ -0,0 +1,439 @@
+    }
+}
+
+#[derive(Clone)]
+pub struct WriterInactive {}
+
+impl WriterLock for WriterInactive {
+    #[cfg(test)]
+    fn tell_me_who_you_are(&self) -> String {
+        "WriterInactive".to_string()
+    }
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+pub trait InterfaceRaw {
+    fn read_line(&self) -> std::io::Result<ReadResult>;
+    fn add_history_unique(&self, line: String);
+    fn lock_writer_append(&self) -> std::io::Result<Box<dyn WriterLock + '_>>;
+    fn set_prompt(&self, prompt: &str) -> std::io::Result<()>;
+}
+
+impl<U: linefeed::Terminal + 'static> InterfaceRaw for Interface<U> {
+    fn read_line(&self) -> std::io::Result<ReadResult> {
+        self.read_line()
+    }
+
+    fn add_history_unique(&self, line: String) {
+        self.add_history_unique(line);
+    }
+
+    fn lock_writer_append(&self) -> std::io::Result<Box<dyn WriterLock + '_>> {
+        match self.lock_writer_append() {
+            Ok(writer) => Ok(Box::new(writer)),
+            //untested ...mocking here would require own definition of a terminal type, I saw how
+            //that's done like that and it takes many objects to be implemented because of the nature
+            //of that external library; I think it isn't worth it
+            Err(error) => Err(error),
+        }
+    }
+
+    fn set_prompt(&self, prompt: &str) -> std::io::Result<()> {
+        self.set_prompt(prompt)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::test_utils::mocks::{InterfaceRawMock, MixingStdout, TerminalActiveMock};
+    use crossbeam_channel::unbounded;
+    use linefeed::DefaultTerminal;
+    use std::io::{Error, Write};
+    use std::sync::Barrier;
+    use std::thread;
+    use std::time::Duration;
+
+    //In the two following tests I use the system stdout handles, which is the standard way in the project, but thanks to
+    //the lock provided by TerminalWrapper, it'll protect one from any influence of another.
+
+    #[test]
+    fn terminal_wrapper_without_lock_does_not_block_others_from_writing_into_stdout() {
+        let closure1: Box<dyn FnMut(TerminalWrapper, MixingStdout) + Sync + Send> =
+            Box::new(move |_interface: TerminalWrapper, mut stdout_c| {
+                write_in_cycles("AAA", &mut stdout_c);
+            });
+
+        let closure2: Box<dyn FnMut(TerminalWrapper, MixingStdout) + Sync + Send> =
+            Box::new(move |_interface: TerminalWrapper, mut stdout_c| {
+                write_in_cycles("BBB", &mut stdout_c);
+            });
+
+        let given_output = test_terminal_collision(Box::new(closure1), Box::new(closure2));
+
+        //in an extreme case it may be printed like one group is complete and the other is divided
+        let results = [
+            given_output.contains(&"A".repeat(90)),
+            given_output.contains(&"B".repeat(90)),
+        ];
+        let at_least_one = results.iter().find(|bool| **bool == false);
```

Comment:
Investigate the any() method: it may do what you want more cleanly. Also: don't name a variable "bool."